core-services/src/main/java/org/zalando/nakadi/service/publishing/AvroEventPublisher.java

```diff
@@ -0,0 +1,155 @@
+            metadata.put("eid", uuidGenerator.randomUUID().toString());
+            metadata.put("flow_id", FlowIdUtils.peek());
+            metadata.put("event_type", etName);
+            metadata.put("partition", 0); // fixme avro
+            metadata.put("received_at", new DateTime(DateTimeZone.UTC));
+            metadata.put("schema_version", "0");  // fixme avro
+            metadata.put("published_by", publishedBy);
+
+            Closeable publishingCloser = null;
+            try {
+                publishingCloser = timelineSync.workWithEventType(etName, nakadiSettings.getTimelineWaitTimeoutMs());
+
+                // publish under timeline lock
+                serializeAndSendToKafka(etName, metadata, event);
+
+            } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+                LOG.error("Failed to wait for timeline switch", e);
+            } catch (final TimeoutException e) {
+                LOG.error("Failed to wait for timeline switch", e);
+            } finally {
+                try {
+                    if (publishingCloser != null) {
+                        publishingCloser.close();
+                    }
+                } catch (final IOException e) {
+                    LOG.error("Exception occurred when releasing usage of event-type", e);
+                }
+            }
+
+        } catch (final Exception e) {
+            LOG.error("Error occurred when submitting KPI event for publishing", e);
+        }
+    }
+
+    private void serializeAndSendToKafka(final String eventTypeName,
+                                         final GenericRecord eventMetadata,
+                                         final GenericRecord event) {
+        final EventType eventType = eventTypeCache.getEventType(eventTypeName);
+        final Timeline activeTimeline = timelineService.getActiveTimeline(eventType);
+        final String topic = activeTimeline.getTopic();
+
+        final Span publishingSpan = TracingService.buildNewSpan("publishing_to_kafka")
+                .withTag(Tags.MESSAGE_BUS_DESTINATION.getKey(), topic)
+                .withTag("event_type", eventTypeName)
+                .withTag("type", "avro")
+                .start();
+
+        try (Closeable ignored = TracingService.activateSpan(publishingSpan)) {
+            final ByteArrayOutputStream eventOutputStream = new ByteArrayOutputStream();
+            final GenericDatumWriter genericDatumWriter = new GenericDatumWriter(eventMetadata.getSchema());
+            genericDatumWriter.setSchema(event.getSchema());
+            genericDatumWriter.write(event, EncoderFactory.get()
+                    .directBinaryEncoder(eventOutputStream, null));
+            final byte[] metadata = eventOutputStream.toByteArray();
+
+            genericDatumWriter.setSchema(event.getSchema());
+            genericDatumWriter.write(event, EncoderFactory.get()
```

Comment:
I'm not sure that this is correct use of the output stream.  I'd sleep better if we've taken a separate instance here ;)

Reply:
I do not see issue with reconfiguration

```
  public void setSchema(Schema root) {
    this.root = root;
  }

  public void write(D datum, Encoder out) throws IOException {
    Objects.requireNonNull(out, "Encoder cannot be null");
    write(root, datum, out);
  }
```

for your sleep Alex, I will create a new writer :)